### PR TITLE
[RHSTOR-4798] MDSCacheUsageHigh alert feature test cases automation.

### PR DESCRIPTION
**Pull request by nagendra202**
[RHSTOR-4798] MDSCacheUsageHigh alert feature test cases automation.

Feature Released: 4.15

Test Plan: https://docs.google.com/spreadsheets/d/172CiopML3LbPC3AN9Dm6lKFzfvbs-Etxj6YTYPBZ9Ew/edit#gid=1005735948

Automated below two tests: [linked to Polarion OCS-5478]
1. [RHSTOR-4798](https://issues.redhat.com//browse/RHSTOR-4798) Verify MDSCacheUsageHigh alert and actionable link.
2. [RHSTOR-4798](https://issues.redhat.com//browse/RHSTOR-4798) Verify alert with node drain.

### CODE REVIEW

File: kubeconfig [hunk 1]
```diff
@@ -0,0 +1,52 @@
+apiVersion: v1
```
Comment (fbalak) [inline kubeconfig:1]: This should not be part of the pr

**Review by fbalak — CHANGES_REQUESTED** [review]

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cpu_high_usage.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+from concurrent.futures import ThreadPoolExecutor
+from ocs_ci.helpers import helpers
+from ocs_ci.ocs import cluster
+from ocs_ci.utility import prometheus
+from ocs_ci.framework.pytest_customization.marks import magenta_squad
+from ocs_ci.utility.utils import ceph_health_check_base
+
+log = logging.getLogger(__name__)
+
+
```
Comment (DanielOsypenko) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cpu_high_usage.py:16]: we miss tier, polarion-id, and squad decorators

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+from ocs_ci.ocs.node import (
+    unschedule_nodes,
+    drain_nodes,
+    schedule_nodes,
+)
+
+log = logging.getLogger(__name__)
+
+
+@pytest.fixture(scope="function")
```
Comment (DanielOsypenko) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:22]: we miss tier, polarion-id and squad decorators

File: ocs_ci/templates/workloads/helper_scripts/meta_data_io.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            )
+
+            # Release the lock
+            fcntl.flock(lock_file, fcntl.LOCK_UN)
+
+    except Exception as e:
+        logging.error(f"Error during metadata operations: {str(e)}")
+
+
+if __name__ == "__main__":
```
Comment (OdedViner) [inline ocs_ci/templates/workloads/helper_scripts/meta_data_io.py:100]: Would you like to run this page directly? if yes, why?

File: kubeconfig (hunk 1 above)
Comment (hnallurv) [reply kubeconfig:1]: @nagendra202 please check and address the above comment at the earliest

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 3 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:22]: Added tier and squad name just above the class. Polarion id added for all tests.

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cpu_high_usage.py (hunk 2 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cpu_high_usage.py:16]: Added tier and squad name just above the class. Polarion id added for all tests.

File: kubeconfig (hunk 1 above)
Comment (nagendra202) [reply kubeconfig]: Not sure how this file has been pushed. Anyhow deleted it.

File: ocs_ci/templates/workloads/helper_scripts/meta_data_io.py (hunk 4 above)
Comment (nagendra202) [reply ocs_ci/templates/workloads/helper_scripts/meta_data_io.py:100]: This is a standalone script will run inside the fedora pods for few mins to generate load in cache.

File: ocs_ci/templates/workloads/helper_scripts/file_creator_io.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    with ThreadPoolExecutor(max_workers=num_clients) as executor:
+        futures = [
+            executor.submit(create_files, base_path, num_files)
+            for _ in range(num_clients)
+        ]
+        results = [future.result() for future in futures]
+        return results
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (nagendra202) [reply ocs_ci/templates/workloads/helper_scripts/file_creator_io.py:31]: This is a standalone script will run inside the fedora pods for few mins to generate load on CPU.

File: kubeconfig (hunk 1 above)
Comment (hnallurv) [reply kubeconfig]: @fbalak could you please review the rest and provide your comments?

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        # Copy meta_data_io.py to fedora pod
+        log.info("Copying meta_data_io.py to fedora pod ")
+        cmd = f"oc cp {file} {pvc_obj.namespace}/{pod_obj.name}:/"
+        helpers.run_cmd(cmd=cmd)
+        log.info("meta_data_io.py copied successfully ")
+
+        # Run meta_data_io.py on fedora pod
+        log.info("Running meta data IO on fedora pod ")
+        metaio_executor = ThreadPoolExecutor(max_workers=1)
+        # self.metaio_thread = metaio_executor.submit(
```
Comment (DanielOsypenko) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:55]: comment left

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+    @pytest.mark.polarion_id("OCS-5571")
+    def test_mds_cache_alert_with_active_node_drain(
+        self, run_metadata_io_with_cephfs, threading_lock
+    ):
+        # This function verifies the mds cache alert with active mds running node drain
+        log.info(
+            "Metadata IO started in the background. Lets wait for 15 minutes before validating the MDS alert"
+        )
+        time.sleep(900)
```
Comment (DanielOsypenko) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:117]: Not critical.

do you want to pull these timeouts to module or class variables? 
I see in PR 900, 600, 100 sec. This may help to maintain the module.

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+        time.sleep(900)
+        assert active_mds_alert_values(threading_lock)
+
+    @pytest.mark.polarion_id("OCS-5579")
+    def test_mds_cache_alert_with_all_mds_node_scaledown(
+        self, run_metadata_io_with_cephfs, threading_lock
+    ):
+        # This function verifies the mds cache alert with both active and standby-replay mds scale down and up
+        time.sleep(900)
```
Comment (DanielOsypenko) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:187]: Please add logger.info message before every time.sleep that we are doing, so we can identify timeouts in the log console, including the sleep period. Same as you did at [tests/functional/monitoring/prometheus/alerts/test_alert_mds_cpu_high_usage.py](https://github.com/red-hat-storage/ocs-ci/pull/9757/files#diff-092955c44cfe901fa9242fe90704a92a42b7dcdabb650727da505351ac4a2cea)

 ```
log.info(
            "File creation IO started in the background."
            " Script will sleep for 15 minutes before validating the MDS alert"
        )

```

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 6 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py]: Removed the commented line of code.

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 7 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py]: Good suggestion, implemented the same.

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 8 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py]: Implemented the changes.

File: ocs_ci/ocs/cluster.py [hunk 9]
```diff
@@ -3316,3 +3316,68 @@ def check_cephcluster_status(
...
+        including the following keys in case of success, otherwise None.
+        - "node_ip": The IP address of the node running the active MDS daemon.
+        - "mds_daemon": The name of the MDS daemon.
+        - "active_pod": The name of the active pod.
+        - "node_name": The name of the node where active mds pod is running.
+    """
+    ct_pod = pod.get_ceph_tools_pod()
+    ceph_mdsmap = ct_pod.exec_ceph_cmd("ceph fs status")
+
+    # Find ceph daemon state as 'active'
```
Comment (PrasadDesala) [inline ocs_ci/ocs/cluster.py:3335]: you can remove the commented code

File: ocs_ci/ocs/cluster.py [hunk 10]
```diff
@@ -3316,3 +3316,68 @@ def check_cephcluster_status(
...
+        None,
+    )
+
+    if ceph_daemon_name is None:
+        logger.error("No active MDS daemon found")
+        return None
+
+    logger.info(f"Found active MDS daemon: {ceph_daemon_name}")
+
+    # Find ceph MDS pod running 'active' daemon.
```
Comment (PrasadDesala) [inline ocs_ci/ocs/cluster.py:3351]: all these comments can be removed, instead convert it to log message

File: ocs_ci/ocs/cluster.py [hunk 11]
```diff
@@ -3316,3 +3316,68 @@ def check_cephcluster_status(
...
+    mds_pods = get_mds_pods()
+    active_pod = next((srp for srp in mds_pods if ceph_daemon_name in srp.name), None)
```
Comment (PrasadDesala) [inline ocs_ci/ocs/cluster.py:3353]: replace srp with some meaningful name, may be pod?

File: ocs_ci/ocs/cluster.py [hunk 12]
```diff
@@ -3316,3 +3316,68 @@ def check_cephcluster_status(
...
+
+    # Get the node IP of running pod
+    node_ip = active_pod.data["status"].get("hostIP")
+    if not node_ip:
+        logger.error(
+            f"Unable to determine IP address of node running active MDS pod '{active_pod.name}'"
+        )
+        return None
+    # Get the node name of running pod.
+    node_name = active_pod.data["spec"].get("nodeName")
```
Comment (PrasadDesala) [inline ocs_ci/ocs/cluster.py:3371]: use .get("spec") to prevent potential keyerror

File: ocs_ci/ocs/constants.py [hunk 13]
```diff
@@ -2862,3 +2865,10 @@
 
 # ascii characters
 ENTER_KEY = chr(13)
+
+# helper script to perform meta data operations [create, rename, change permissions...etc] on app pod to fill MDS cache
+METAIO = os.path.join(TEMPLATE_WORKLOAD_DIR, "helper_scripts/meta_data_io.py")
+# helper script to perform file creation IO on app pod to fill MDS cpu
+FILE_CREATOR_IO = os.path.join(
```
Comment (PrasadDesala) [inline ocs_ci/ocs/constants.py:2872]: Not a blocker for this PR, but IMO we should push this script to an image and run it as a container

File: ocs_ci/utility/prometheus.py [hunk 14]
```diff
@@ -783,3 +794,75 @@ def unsubscribe(self):
...
+        alert_name (str): Alert name
+        msg (str): Alert message
+        alerts (list): List of alerts to check
+        state (str): state of the alert
+        active_mds(str): active mds daemon name
+        standby_mds(str): standby-replay mds daemon name
+        description(str): description of the alert triggered
+        runbook(str): runbook_url in the alert
+        severity(str): severity of the alert
+    """
```
Comment (PrasadDesala) [inline ocs_ci/utility/prometheus.py:824]: add an empty line above this

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+        standby_mds=sr_mds,
+    )
+    log.info("Alert verified successfully")
+    return True
+
+
+@tier2
+@blue_squad
+class TestMdsMemoryAlerts:
+    alert_timer = 900  # sleep time to generate the alert 15 minutes
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:98]: add an empty line above this

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+    scale_timer = 100  # sleep time to wait after running scale down
+
+    @pytest.mark.polarion_id("OCS-5570")
+    def test_alert_triggered(
+        self, run_metadata_io_with_cephfs, threading_lock, alert_timer
+    ):
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:104]: Please add a description about the test case

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+        )
+        time.sleep(alert_timer)
+        log.info("Validating the alert now")
+        assert active_mds_alert_values(threading_lock)
+
+    @pytest.mark.polarion_id("OCS-5571")
+    def test_mds_cache_alert_with_active_node_drain(
+        self, run_metadata_io_with_cephfs, threading_lock, alert_timer
+    ):
+        # This function verifies the mds cache alert with active mds running node drain
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:117]: Please add a description about the test case

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+        log.info(
+            "Metadata IO started in the background. Lets wait for 15 minutes before validating the MDS alert"
+        )
+        time.sleep(alert_timer)
+        log.info("Validating the alert now")
+        assert active_mds_alert_values(threading_lock)
+
+        node_name = cluster.get_active_mds_info()["node_name"]
+        # Unschedule active mds running node.
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:126]: you can remove the comments at all the places since they are repetitive

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+    3. Copy helper_scripts/meta_data_io.py to Fedora dc pod
+    4. Run meta_data_io.py on fedora pod
+    """
+    access_mode = constants.ACCESS_MODE_RWX
+    file = constants.METAIO
+    interface = constants.CEPHFILESYSTEM
+    log.info("Checking for Ceph Health OK")
+    ceph_health_check_base()
+
+    for i in range(3):
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:37]: replace i with some meaningful name

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+
+@pytest.fixture(scope="function")
+def run_metadata_io_with_cephfs(pvc_factory, dc_pod_factory):
+    """
+    This function facilitates
+    1. Create PVC with Cephfs, access mode RWX
+    2. Create dc pod with Fedora image
+    3. Copy helper_scripts/meta_data_io.py to Fedora dc pod
+    4. Run meta_data_io.py on fedora pod
+    """
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:30]: add an empty line above

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+    access_mode = constants.ACCESS_MODE_RWX
+    file = constants.METAIO
+    interface = constants.CEPHFILESYSTEM
+    log.info("Checking for Ceph Health OK")
+    ceph_health_check_base()
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:35]: this check is already done at the beginning of the test execution by our framework, I think we can remove it here

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 22]
```diff
@@ -0,0 +1,234 @@
...
+
+    for i in range(3):
+        # Creating PVC with cephfs as interface
+        log.info(f"Creating {interface} based PVC")
+        pvc_obj = pvc_factory(interface=interface, access_mode=access_mode, size="30")
+        # Creating a Fedora dc pod
+        log.info("Creating fedora dc pod")
+        pod_obj = dc_pod_factory(
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:43]: dc_pod_factory can even create a PVC, is there any reason why you are using PVC separately using pvc_factory above?

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+            pvc=pvc_obj, access_mode=access_mode, interface=interface
+        )
+        # Copy meta_data_io.py to fedora pod
+        log.info("Copying meta_data_io.py to fedora pod ")
+        cmd = f"oc cp {file} {pvc_obj.namespace}/{pod_obj.name}:/"
+        helpers.run_cmd(cmd=cmd)
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:49]: since you have the pod_obj, you could use copy_to_pod_rsync func to copy the file from a local path to the pod path

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+
+        # Run meta_data_io.py on fedora pod
+        log.info("Running meta data IO on fedora pod ")
+        metaio_executor = ThreadPoolExecutor(max_workers=1)
+        metaio_executor.submit(
+            pod_obj.exec_sh_cmd_on_pod, command="python3 meta_data_io.py"
+        )
+
+
+def active_mds_alert_values(threading_lock):
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:60]: doc string missing

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+    active_mds = cluster.get_active_mds_info()["mds_daemon"]
+    sr_mds = cluster.get_mds_standby_replay_info()["mds_daemon"]
+    cache_alert = constants.ALERT_MDSCACHEUSAGEHIGH
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:63]: since you are calling it only once, you can directly call the constant in the function

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 26]
```diff
@@ -0,0 +1,234 @@
...
+    description = (
+        f"MDS cache usage for the daemon mds.{active_mds} has exceeded above 95% of the requested value."
+        f" Increase the memory request for mds.{active_mds} pod."
+    )
+    runbook = (
+        "https://github.com/openshift/runbooks/blob/master/alerts/"
+        "openshift-container-storage-operator/CephMdsCacheUsageHigh.md"
+    )
+    state = "firing"
+    severity = "critical"
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:74]: since you are calling it only once, you can directly call the constant in the function

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 27]
```diff
@@ -0,0 +1,234 @@
...
+
+        node_name = cluster.get_active_mds_info()["node_name"]
+        # Unschedule active mds running node.
+        unschedule_nodes([node_name])
+        log.info(f"node {node_name} unscheduled successfully")
+        # Drain node operation
+        drain_nodes([node_name])
+        log.info(f"node {node_name} drained successfully")
+        # Make the node schedule-able
+        schedule_nodes([node_name])
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:133]: how do you make sure that, the app pod running node is not picked here?

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py [hunk 28]
```diff
@@ -0,0 +1,234 @@
...
+        log.info(
+            " Script will be in sleep for 100 seconds to make sure active mds scale down completed."
+        )
+        time.sleep(scale_timer)
+        log.info(f"Scale up {deployment_name} to 1")
+        helpers.modify_deployment_replica_count(
+            deployment_name=deployment_name, replica_count=1
+        )
+        log.info(
+            "Metadata IO started in the background. Script will sleep for 15 minutes before validating the MDS alert"
```
Comment (PrasadDesala) [inline tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:165]: how do you make sure that, the app pod running node is not picked here?

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 27 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:133]: We need to bring down a node where Active MDS pod is running. So the same  node name given for node drain, the same node name is giving for schedule_nodes as well.

File: ocs_ci/ocs/cluster.py (hunk 9 above)
Comment (nagendra202) [reply ocs_ci/ocs/cluster.py]: done

File: ocs_ci/ocs/cluster.py (hunk 10 above)
Comment (nagendra202) [reply ocs_ci/ocs/cluster.py]: done

File: ocs_ci/ocs/cluster.py (hunk 11 above)
Comment (nagendra202) [reply ocs_ci/ocs/cluster.py]: done

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 24 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:67]: added doc string

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 25 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:79]: Anyhow I have to import constants for other variables so I just added it to the alerts section where everyone added the alerts in the constants.py. I can hard code it by just giving alert name within this function. I feel not to do that way because in future if someone want that alert to reuse they can refer alerts section in constant.py

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 26 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:90]: Anyhow I have to import constants for other variables so I just added it to the alerts section where everyone added the alerts in the constants.py. I can hard code it by just giving alert name within this function. I feel not to do that way because in future if someone want that alert to reuse they can refer alerts section in constant.py

**Review by nagendra202 — COMMENTED** [review]
Addressed all review comments and pushed changes.

**Review by fbalak — CHANGES_REQUESTED** [review]
Other monitoring tests use concept of measurement fixtures: the system is moved into a disturbed state state in fixture, after that there is started alert collecting in a thread to monitor how the system behaves and then the measurement is saved. This way you can use those measurements in multiple test cases and you have logged history of alerts that come into the system.
You reimplemented this. In your approach you wait for alert in Firing state but you are not checking for Pending state.
Why didn't you reuse existing resources like measure_operation function in combination with fixtures as done in https://github.com/red-hat-storage/ocs-ci/blob/master/tests/functional/monitoring/conftest.py?

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 15 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py]: added an empty line.

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 16 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py]: done

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 17 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py]: done

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 18 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py]: done

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 19 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py]: done

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 20 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:35]: done

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 23 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:58]: To approach that way, it will be adding few more lines of code and need to import Pod class for that. Here only once we copy the file to the pod. So I feel ok to use the single line command for now. If any issues arise then will look into this?

File: ocs_ci/utility/prometheus.py [hunk 29]
```diff
@@ -783,3 +794,75 @@ def unsubscribe(self):
...
+    alerts,
+    active_mds,
+    standby_mds,
+    description,
+    runbook,
+    state,
+    severity,
+):
+    """
+    Check all the properties like severity, state, message, description, runbook_url
```
[the remark this answers is not in the document]
Comment (nagendra202) [reply ocs_ci/utility/prometheus.py]: made changes

File: ocs_ci/utility/prometheus.py [hunk 30]
```diff
@@ -63,6 +63,17 @@ def check_alert_list(
             )
             assert len(found_alerts) == 1, assert_msg
 
+        assert_msg = f"Alert message for alert {label} is not correct"
```
[the remark this answers is not in the document]
Comment (nagendra202) [reply ocs_ci/utility/prometheus.py:75]: made changes

File: ocs_ci/templates/workloads/helper_scripts/meta_data_io.py [hunk 31]
```diff
@@ -0,0 +1,111 @@
+import time
```
[the remark this answers is not in the document]
Comment (nagendra202) [reply ocs_ci/templates/workloads/helper_scripts/meta_data_io.py:6]: done.

**Review by nagendra202 — COMMENTED** [review]
addressed review comments.

File: tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py (hunk 22 above)
Comment (nagendra202) [reply tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:58]: done

**Review by nagendra202 — COMMENTED** [review]
Replied to review comments.

**Comment by nagendra202** [comment]
@OdedViner  and @fbalak  I have addressed all review comments and stabilised code very well now. Please review and provide your feedback.  There were failures in PR job validation as part of teardown. I have taken care and rerunning the PR validation now.

**Review by nagendra202 — COMMENTED** [review]
Addressed all review and stabilised code in tests and teardown. Please review it.

**Review by fbalak — APPROVED** [review]

**Review by Akarsha-rai — APPROVED** [review]
